### PR TITLE
Small Operator Parsing

### PR DESCRIPTION
**Pull request by mddvisu**
Added parsing for the increment, decrement, double, negative, and square small operator nodes.

Graph Screenshot:

image

Translated Lua file:

```

local smalloperatorparsetest_VM = {}

if _G.SCRIPT_CANVAS_GLOBAL_RELEASE then
-- **********************************************************************************************************************
-- ****** release configuration, no debug information available, no performance markers

smalloperatorparsetest_VM.className_SCVM = "smalloperatorparsetest_VM"
function smalloperatorparsetest_VM.OnGraphStart(executionState)
	local m_Testvar_scvm = 6.000000
	local m_Testvar_scvm_output_scvm = m_Testvar_scvm
	m_Testvar_scvm_output_scvm = (m_Testvar_scvm_output_scvm * 2.000000)
	local format_output_scvm = string.format([[%.4f]], m_Testvar_scvm_output_scvm)
	Debug.Log(format_output_scvm)
end

-- ***** end release configuration
-- **********************************************************************************************************************
elseif _G.SCRIPT_CANVAS_GLOBAL_PERFORMANCE then
-- **********************************************************************************************************************
-- ***** performance configuration, no debug information available, performance markers in place

smalloperatorparsetest_VM.className_SCVM = "smalloperatorparsetest_VM"
function smalloperatorparsetest_VM.OnGraphStart(executionState)
	local m_Testvar_scvm = 6.000000
	local m_Testvar_scvm_output_scvm = m_Testvar_scvm
	m_Testvar_scvm_output_scvm = (m_Testvar_scvm_output_scvm * 2.000000)
	local format_output_scvm = string.format([[%.4f]], m_Testvar_scvm_output_scvm)
	Debug.Log(format_output_scvm)
end

-- ***** end performance configuration
-- **********************************************************************************************************************
else
-- **********************************************************************************************************************
-- ***** debug configuration, debug information available upon when tracing is requested, no performance markers in place

smalloperatorparsetest_VM.className_SCVM = "smalloperatorparsetest_VM"
function smalloperatorparsetest_VM.OnGraphStart(executionState)
	if DebugIsTraced(executionState) then
		DEBUG_SIGNAL_OUT_SUBGRAPH(executionState, '{51D2B631-90FD-56FF-81AE-733C8CD157A3}:163310ae', 0) -- TranslateFunctionBlock begin
		local m_Testvar_scvm = 6.000000
		DEBUG_SIGNAL_IN_SUBGRAPH(executionState, '{51D2B631-90FD-56FF-81AE-733C8CD157A3}:163310ae', 0, m_Testvar_scvm) -- TranslateExecutionTreeFunctionCall begin
		local m_Testvar_scvm_output_scvm = m_Testvar_scvm
		DEBUG_SIGNAL_OUT_SUBGRAPH(executionState, '{51D2B631-90FD-56FF-81AE-733C8CD157A3}:163310ae', 1, m_Testvar_scvm_output_scvm) -- TranslateExecutionTreeFunctionCall end
		m_Testvar_scvm_output_scvm = (m_Testvar_scvm_output_scvm * 2.000000)
		DEBUG_SIGNAL_IN_SUBGRAPH(executionState, '{51D2B631-90FD-56FF-81AE-733C8CD157A3}:163310ae', 1, [[%.4f]], m_Testvar_scvm_output_scvm) -- TranslateExecutionTreeFunctionCall begin
		local format_output_scvm = string.format([[%.4f]], m_Testvar_scvm_output_scvm)
		DEBUG_SIGNAL_OUT_SUBGRAPH(executionState, '{51D2B631-90FD-56FF-81AE-733C8CD157A3}:163310ae', 2, format_output_scvm) -- TranslateExecutionTreeFunctionCall end
		DEBUG_SIGNAL_IN_SUBGRAPH(executionState, '{51D2B631-90FD-56FF-81AE-733C8CD157A3}:163310ae', 2, format_output_scvm) -- TranslateExecutionTreeFunctionCall begin
		Debug.Log(format_output_scvm)
	else
		local m_Testvar_scvm = 6.000000
		local m_Testvar_scvm_output_scvm = m_Testvar_scvm
		m_Testvar_scvm_output_scvm = (m_Testvar_scvm_output_scvm * 2.000000)
		local format_output_scvm = string.format([[%.4f]], m_Testvar_scvm_output_scvm)
		Debug.Log(format_output_scvm)
	end
end

-- ***** end debug configuration
-- **********************************************************************************************************************
end

return smalloperatorparsetest_VM
```

### CODE REVIEW

File: Gems/ScriptCanvas/Code/Include/ScriptCanvas/Grammar/AbstractCodeModel.cpp [hunk 1]
```diff
@@ -4387,10 +4423,104 @@ namespace ScriptCanvas
...
-                if (auto sourceVariable = ParseConnectedInputData(input, execution, nodes, FirstNode::Parent))
+                if (auto sourceVariableOperationData = ParseConnectedInputData(input, execution, nodes, FirstNode::Parent))
                 {
-                    execution->AddInput({ &input, sourceVariable, DebugDataSource::FromOtherSlot(input.GetId(), input.GetDataType(), sourceVariable->m_sourceSlotId) });
-                    CheckConversion(execution->ModConversions(), sourceVariable, execution->GetInputCount() - 1, input.GetDataType());
+                    if (!sourceVariableOperationData->m_smallOperations.empty())
+                    {
+                        Node* node = aznew Node();
+
+                        ExecutionSlotConfiguration inexconf;
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Include/ScriptCanvas/Grammar/AbstractCodeModel.cpp:4432]: What's the motivation for the new Node here?

Comment (carlitosan) [reply Gems/ScriptCanvas/Code/Include/ScriptCanvas/Grammar/AbstractCodeModel.cpp:4432]: It seems to be unused and leaked...

Comment (mddvisu) [reply Gems/ScriptCanvas/Code/Include/ScriptCanvas/Grammar/AbstractCodeModel.cpp:4432]: This node gets added to the execution path to actually preform the operation.

File: Gems/ScriptCanvas/Code/Include/ScriptCanvas/Translation/GraphToLua.cpp [hunk 2]
```diff
@@ -1872,13 +1872,13 @@ namespace ScriptCanvas
                 if (canWriteValue(isformatStingInput, input))
                 {
                     CheckConversion converter(m_dotLua, execution->GetInput(index).m_value, execution->GetConversions(), index);
-                    WriteVariableReference(input);
+                    WriteVariableReference(execution->GetInput(index).m_value);
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Include/ScriptCanvas/Translation/GraphToLua.cpp:1875]: Nice catch!

**Comment by mddvisu** [comment]
_**Known issues:**_

_Issue 1:_

bug1

For the graph above, one would expect the print node to print double the original value of Testvar, as the small operator node is connected to the GetVariable node and not the add node. This is not the case, however, as the output is 12. This is because it is adding 2 to the variable before preforming the operation. If we take a peek at the Lua script:

```
function smalloperatorparsetest_VM.OnGraphStart(executionState)
	local m_Testvar_scvm = 4.000000
	local m_Testvar_scvm_output_scvm = m_Testvar_scvm
	m_Testvar_scvm_output_scvm = (m_Testvar_scvm_output_scvm + 2.000000)
	m_Testvar_scvm_output_scvm = (m_Testvar_scvm_output_scvm * 2.000000)
	local format_output_scvm = string.format([[%.4f]], m_Testvar_scvm_output_scvm)
	Debug.Log(format_output_scvm)
end
```

_Issue 2:_

We can see that it is using the same variable the Add node uses to multiply it by two after adding two.

bug2

The graph above gives errors, even though the Invert small operator is implemented. This very likely has to do with the extra property extraction slots attached to get vector 2 and how it interacts with the small operator. The error being thrown is in ScriptCanvas::Grammar::AbstractCodeModel::ParsePropertyExtractionsPost.

**Comment by mddvisu** [comment]
Closed in favor of https://github.com/o3de/o3de/pull/16382, https://github.com/o3de/o3de/pull/16403, https://github.com/o3de/o3de/pull/16412, and https://github.com/o3de/o3de/pull/16404

**Pull request closed without merging**